### PR TITLE
[CLI] Invoke tests

### LINKED ISSUE #2111: Add tests for invoke
Add e2e tests for invoke command

### PR DESCRIPTION
**Pull request by war-in**
### Checklist
- [x] Relevant issue is linked
- [ ] Docs updated/issue for docs created
- [x] Added relevant tests

Closes #2111

### CODE REVIEW

**Comment by karol-bisztyga** [comment]
This should target the master branch

**Review by karol-bisztyga — CHANGES_REQUESTED** [review]
This should target the master branch

File: cast/tests/e2e/helpers/fixtures.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
+use crate::helpers::constants::{
```
Comment (karol-bisztyga) [inline cast/tests/e2e/helpers/fixtures.rs:1]: the same as in https://github.com/software-mansion/protostar/pull/2150#discussion_r1250601906, I don't think these are e2e tests, or are they?

File: cast/tests/e2e/helpers/fixtures.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        ACCOUNT,
+        &Utf8PathBuf::from(ACCOUNT_FILE_PATH),
+        provider,
+        &get_network(NETWORK).unwrap(),
+    )
+    .unwrap()
+}
+
+pub async fn declare_simple_balance_contract() {
+    let provider = get_provider(URL).unwrap();
```
Comment (karol-bisztyga) [inline cast/tests/e2e/helpers/fixtures.rs:28]: I do not think we should be using `unwrap` as much as you are, even in tests. Can you replace them at least with `expect`? Even this tiny error message can sometimes be helpful, you don't need to dig into the code, you sometimes just see the message and know what is wrong right away.

Comment (karol-bisztyga) [reply cast/tests/e2e/helpers/fixtures.rs:28]: What happens when a test panics? All tests break or just one fails?

**Review by karol-bisztyga — CHANGES_REQUESTED** [review]

File: cast/tests/e2e/helpers/fixtures.rs (hunk 1 above)
Comment (THenry14) [reply cast/tests/e2e/helpers/fixtures.rs:1]: these are not tests

File: cast/tests/e2e/helpers/fixtures.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
+use crate::helpers::constants::{
+    ACCOUNT, ACCOUNT_FILE_PATH, CASM_CONTRACT_PATH, NETWORK, SIERRA_CONTRACT_PATH, URL,
+};
+use camino::Utf8PathBuf;
+use cast::{get_account, get_network, get_provider};
+use starknet::accounts::{Account, SingleOwnerAccount};
```
Comment (THenry14) [inline cast/tests/e2e/helpers/fixtures.rs:6]: did you try to construct raw requests? were there any problems?
I'm just afraid that we may be overly relying on starknet-rs, and wonder if we could somehow mitigate that

File: cast/tests/e2e/helpers/fixtures.rs (hunk 1 above)
Comment (karol-bisztyga) [reply cast/tests/e2e/helpers/fixtures.rs:1]: right, sorry wrong file

File: cast/tests/e2e/helpers/fixtures.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            "0x8448a68b5ea1affc45e3fd4b8b480ea36a51dc34e337a16d2567d32d0c6f8a",
+        )
+        .unwrap(),
+        account,
+    );
+    let deployment = factory.deploy(Vec::new(), FieldElement::from_hex_be("0x1").unwrap(), false);
+    deployment.send().await.ok();
+}
+
+pub fn common_cli_args() -> Vec<&'static str> {
```
Comment (THenry14) [inline cast/tests/e2e/helpers/fixtures.rs:64]: maybe `default_cli_args`? or something like this?

File: cast/tests/e2e/helpers/fixtures.rs (hunk 3 above)
Comment (war-in) [reply cast/tests/e2e/helpers/fixtures.rs:4]: I've tried but there is still an issue with signing. We have to have an account to sign those transactions :/

**Comment by war-in** [comment]
Please take a look at the new contract data setup @karol-bisztyga @THenry14 🙏

File: cast/tests/e2e/helpers/fixtures.rs (hunk 2 above)
Comment (war-in) [reply cast/tests/e2e/helpers/fixtures.rs:34]: afaik just one

Comment (war-in) [reply cast/tests/e2e/helpers/fixtures.rs]: https://github.com/software-mansion/protostar/pull/2160/commits/d7c616450e7a4dc87ba63e50d23bc1f1ae4516de

File: cast/tests/data/contracts/balance/src/lib.cairo [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#[contract]
```
Comment (THenry14) [inline cast/tests/data/contracts/balance/src/lib.cairo:1]: It's a nitpick, but is the contract's name somehow imposed by scarb or something? if not, let's maybe rename it to something more meaningful?

File: cast/tests/utils/prepare_for_tests.sh [hunk 6]
```diff
@@ -5,20 +5,43 @@ if ! which starknet-devnet > /dev/null 2>&1; then
...
-    pushd "$COMPILER_DIRECTORY"
-    tar -xzvf "$COMPILER_DIRECTORY/release-x86_64-unknown-linux-musl.tar.gz" cairo/bin/starknet-sierra-compile || exit 1
+    wget "$CAIRO_REPO/$COMPILER_VERSION/scarb-$SCARB_VERSION-x86_64-unknown-linux-musl.tar.gz" -P "$DIRECTORY" || exit 1
+    pushd "$DIRECTORY"
+    tar -xzvf "$DIRECTORY/scarb-$SCARB_VERSION-x86_64-unknown-linux-musl.tar.gz" scarb-$SCARB_VERSION-aarch64-apple-darwin/bin/scarb || exit 1
     popd
   fi
 fi
+
+pushd "$DIRECTORY/../data/contracts/balance"
```
Comment (THenry14) [inline cast/tests/utils/prepare_for_tests.sh:45]: let's move contracts compilation to another script. 
We should probably call the compile_contract scrtips before each test run (firt check if scripts were compiled, if not do it, otherwise do nothing).

File: cast/tests/utils/prepare_for_tests.sh [hunk 7]
```diff
@@ -5,20 +5,43 @@ if ! which starknet-devnet > /dev/null 2>&1; then
   exit 1
 fi
 
-COMPILER_DIRECTORY="$(git rev-parse --show-toplevel)/cast/tests/utils/"
+DIRECTORY="$(git rev-parse --show-toplevel)/cast/tests/utils/"
```
Comment (THenry14) [inline cast/tests/utils/prepare_for_tests.sh:8]:
```suggestion
UTILS_DIRECTORY="$(git rev-parse --show-toplevel)/cast/tests/utils/"
```

File: cast/tests/utils/prepare_for_tests.sh [hunk 8]
```diff
@@ -5,20 +5,43 @@ if ! which starknet-devnet > /dev/null 2>&1; then
...
+
+  elif [[ $(uname -s) == 'Linux' ]]; then
+    wget "$CAIRO_REPO/$COMPILER_VERSION/release-x86_64-unknown-linux-musl.tar.gz" -P "$DIRECTORY" || exit 1
+    pushd "$DIRECTORY"
+    tar -xzvf "$DIRECTORY/release-x86_64-unknown-linux-musl.tar.gz" cairo/bin/starknet-sierra-compile || exit 1
+    popd
+  fi
+fi
+
+if [ ! -x "$COMPILER_DIRECTORY/scarb*/bin/scarb" ]; then
```
Comment (THenry14) [inline cast/tests/utils/prepare_for_tests.sh:30]: 1. can scarb do multiple versions? if so, we could just check if there's required scarb version in system
2. if not, let's get move `if` statement to separate function and reuse it

**Review by karol-bisztyga — CHANGES_REQUESTED** [review]

**Comment by war-in** [comment]
`[Disclaimer]` Many changes are here because the `master` branch was not correctly formatted

**Comment by karol-bisztyga** [comment]
> `[Disclaimer]` Many changes are here because the `master` branch was not correctly formatted

I do not think this is good. I think there should be another PR for formatting and then this PR (invoke tests) should be rebased to the fmt PR.

**Comment by war-in** [comment]
> I do not think this is good. I think there should be another PR for formatting and then this PR (invoke tests) should be rebased to the fmt PR.
Agree, but we should create some checks for formatting to avoid such a situation in the future

**Comment by THenry14** [comment]
> Agree, but we should create some checks for formatting to avoid such a situation in the future

Next in pipeline #2114

File: cast/tests/utils/build_contracts.sh [hunk 9]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+CONTRACTS_DIRECTORY="$(git rev-parse --show-toplevel)/cast/tests/data/contracts"
+
+if command -v scarb &> /dev/null; then
+  for contract_dir in "$CONTRACTS_DIRECTORY"/*; do
+    if ! test -d "$contract_dir"/target; then
+      pushd $contract_dir
```
Comment (karol-bisztyga) [inline cast/tests/utils/build_contracts.sh:8]:
```suggestion
      pushd "$contract_dir"
```

**Review by karol-bisztyga — CHANGES_REQUESTED** [review]

File: cast/tests/data/contracts/balance/src/lib.cairo (hunk 5 above)
Comment (THenry14) [reply cast/tests/data/contracts/balance/src/lib.cairo:1]: by contract name I meant both the name ( ( ͡° ͜ʖ ͡°) ) and the file name

File: cast/tests/e2e/helpers/devnet.rs [hunk 10]
```diff
@@ -46,6 +45,10 @@ fn start_devnet() {
             std::process::exit(1);
         }
     }
+
+    let rt = Runtime::new().expect("Could not instantiate Runtime");
+    rt.block_on(declare_simple_balance_contract());
+    rt.block_on(deploy_simple_balance_contract());
```
[the remark this answers is not in the document]
Comment (THenry14) [reply cast/tests/e2e/helpers/devnet.rs:51]: I suppose declare/deploy will work on different/seeded contracts, right? in that case, this is fine. otherwise, let's adjust to Karol's comment

File: cast/tests/e2e/helpers/fixtures.rs [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    provider: &JsonRpcClient<HttpTransport>,
+) -> SingleOwnerAccount<&JsonRpcClient<HttpTransport>, LocalWallet> {
+    get_account(
+        ACCOUNT,
+        &Utf8PathBuf::from(ACCOUNT_FILE_PATH),
+        provider,
+        &get_network(NETWORK).expect("Could not get the network"),
+    )
+    .expect("Could not get the account")
+}
```
Comment (THenry14) [inline cast/tests/e2e/helpers/fixtures.rs:31]: what's the purpose of this?

File: cast/tests/e2e/helpers/fixtures.rs [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        .expect("Could not create FieldElement from hex string"),
+        account,
+    );
+    let deployment = factory.deploy(
+        Vec::new(),
+        FieldElement::from_hex_be("0x1").expect("Could not create FieldElement from hex string"),
+        false,
+    );
+    deployment.send().await.unwrap();
+}
```
Comment (THenry14) [inline cast/tests/e2e/helpers/fixtures.rs:80]: I suppose that those functions could be merged into one? I believe you use them only in one place, one after another right? that way you don't need to hardcode hashes (I suppose), and get cleaner code

File: cast/tests/utils/build_contracts.sh (hunk 9 above)
Comment (THenry14) [reply cast/tests/utils/build_contracts.sh]: actually, let's `""` all the variables

File: cast/tests/utils/prepare_for_tests.sh [hunk 13]
```diff
@@ -5,20 +5,32 @@ if ! which starknet-devnet > /dev/null 2>&1; then
   exit 1
 fi
```
[the remark this answers is not in the document]
Comment (THenry14) [reply cast/tests/utils/prepare_for_tests.sh:8]: > I know this was here already but I guess we can add it here it's just one line

Context matters. You can do `set -e`, but then `|| exit 1` are to be removed, because they are no longer needed. up to you if you prefer one way or another.

File: cast/tests/utils/prepare_for_tests.sh [hunk 14]
```diff
@@ -5,20 +5,32 @@ if ! which starknet-devnet > /dev/null 2>&1; then
...
 
-COMPILER_DIRECTORY="$(git rev-parse --show-toplevel)/cast/tests/utils/"
+UTILS_DIRECTORY="$(git rev-parse --show-toplevel)/cast/tests/utils/"
```
Comment (THenry14) [inline cast/tests/utils/prepare_for_tests.sh:9]: it seems thid directory is no longer needed for other things than compiler? so maybe we can revert this change altogether?

File: cast/tests/e2e/helpers/devnet.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (war-in) [reply cast/tests/e2e/helpers/devnet.rs:51]: Deploy tests are not compromised because we can use different salt for deployment

File: cast/tests/utils/prepare_for_tests.sh [hunk 15]
```diff
@@ -5,20 +5,32 @@ if ! which starknet-devnet > /dev/null 2>&1; then
...
   fi
 fi
+
+if command -v asdf &> /dev/null; then
+  asdf plugin add scarb https://github.com/software-mansion/asdf-scarb.git
+  asdf install scarb $SCARB_VERSION
+  asdf global scarb $SCARB_VERSION
+  scarb --version
+else
+  printf "Please install asdf\n https://asdf-vm.com/guide/getting-started.html#_2-download-asdf\n"
```
Comment (THenry14) [inline cast/tests/utils/prepare_for_tests.sh:36]: let's require it at the top of file, similarly to how we require devnet. Moreover, we should also check if scarb in need version is installed, and do it only if it is not.

File: cast/tests/e2e/helpers/devnet.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (war-in) [reply cast/tests/e2e/helpers/devnet.rs:51]: In the case of Declare, we will have to create some script which temporarily modifies the balance contract (eg. adds a character to one of the methods). It will change the class_hash

File: cast/tests/utils/prepare_for_tests.sh (hunk 15 above)
Comment (war-in) [reply cast/tests/utils/prepare_for_tests.sh:36]: If the required scarb version is here it will not do anything. It will just print scarb version

File: cast/tests/e2e/helpers/fixtures.rs [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+use starknet::signers::LocalWallet;
+use std::sync::Arc;
+
+pub fn sierra_balance_path() -> String {
+    CONTRACTS_DIR.to_string() + "/balance/target/dev/balance_SimpleBalance.sierra.json"
+}
+
+pub fn casm_balance_path() -> String {
+    CONTRACTS_DIR.to_string() + "/balance/target/dev/balance_SimpleBalance.casm.json"
+}
```
Comment (THenry14) [inline cast/tests/e2e/helpers/fixtures.rs:19]: this is unnecessary probably, right?

File: cast/tests/utils/prepare_for_tests.sh [hunk 17]
```diff
@@ -1,24 +1,38 @@
 #!/bin/bash
+set -e
 
 if ! which starknet-devnet > /dev/null 2>&1; then
   echo "starknet-devnet not found, exiting."
   exit 1
 fi
 
+SCARB_VERSION="0.4.1"
```
Comment (THenry14) [inline cast/tests/utils/prepare_for_tests.sh:9]: All variables should be set in one place, preferably at the top

**Review by THenry14 — APPROVED** [review]
🔥

File: cast/tests/utils/prepare_for_tests.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply cast/tests/utils/prepare_for_tests.sh:13]: You are right, I think `set -e` is better (less boilerplate).

**Review by karol-bisztyga — APPROVED** [review]
